pkg/sanity/cleanup.go

```diff
@@ -36,110 +42,284 @@ type VolumeInfo struct {
+}
+
+func (cl *Cleanup) createSnapshot(ctx context.Context, req *csi.CreateSnapshotRequest) (*csi.CreateSnapshotResponse, error) {
+	snap, err := cl.ControllerClient.CreateSnapshot(ctx, req)
+	if err == nil && snap.GetSnapshot().GetSnapshotId() != "" {
+		cl.registerSnapshot(snap.Snapshot.SnapshotId)
+	}
+	return snap, err
+}
+
+func (cl *Cleanup) deleteSnapshot(ctx context.Context, req *csi.DeleteSnapshotRequest) (*csi.DeleteSnapshotResponse, error) {
+	snap, err := cl.ControllerClient.DeleteSnapshot(ctx, req)
+	if err == nil && req.SnapshotId != "" {
+		cl.unregisterSnapshot(req.SnapshotId)
+	}
+	return snap, err
+}
+
+func (cl *Cleanup) registerSnapshot(id string) {
 	cl.mutex.Lock()
 	defer cl.mutex.Unlock()
-	cl.unregisterVolume(name)
+	cl.registerSnapshotNoLock(id)
 }
-func (cl *Cleanup) unregisterVolume(name string) {
-	if cl.volumes != nil {
-		delete(cl.volumes, name)
+
+func (cl *Cleanup) registerSnapshotNoLock(id string) {
+	Expect(id).NotTo(BeEmpty())
+	if cl.snapshots == nil {
+		cl.snapshots = make(map[string]bool)
 	}
+	cl.snapshots[id] = true
 }
 
-// DeleteVolumes stops using the registered volumes and tries to delete all of them.
-func (cl *Cleanup) DeleteVolumes() {
+func (cl *Cleanup) unregisterSnapshot(id string) {
 	cl.mutex.Lock()
 	defer cl.mutex.Unlock()
-	if cl.volumes == nil {
-		return
+	cl.unregisterSnapshotNoLock(id)
+}
+
+func (cl *Cleanup) unregisterSnapshotNoLock(id string) {
+	Expect(id).NotTo(BeEmpty())
+	if cl.snapshots != nil {
+		delete(cl.snapshots, id)
 	}
-	logger := log.New(GinkgoWriter, "cleanup: ", 0)
+}
+
+// Cleanup calls unpublish methods as needed and deletes all volumes and
+// snapshots.
+func (cl *Cleanup) Cleanup() {
+	cl.mutex.Lock()
+	defer cl.mutex.Unlock()
 	ctx := context.Background()
 
-	for name, info := range cl.volumes {
-		logger.Printf("deleting %s = %s", name, info.VolumeID)
-		if _, err := cl.NodeClient.NodeUnpublishVolume(
-			ctx,
-			&csi.NodeUnpublishVolumeRequest{
-				VolumeId:   info.VolumeID,
-				TargetPath: cl.Context.TargetPath + "/target",
-			},
-		); err != nil {
-			logger.Printf("warning: NodeUnpublishVolume: %s", err)
-		}
+	cl.deleteVolumes(ctx)
+	cl.deleteSnapshots(ctx)
+}
+
+func (cl *Cleanup) deleteVolumes(ctx context.Context) {
+	logger := log.New(GinkgoWriter, "cleanup volumes: ", 0)
 
-		if cl.NodeStageSupported {
-			if _, err := cl.NodeClient.NodeUnstageVolume(
+	for volumeID, info := range cl.volumes {
+		logger.Printf("deleting %s", volumeID)
+		if cl.NodeClient != nil {
+			if _, err := cl.NodeUnpublishVolume(
 				ctx,
-				&csi.NodeUnstageVolumeRequest{
-					VolumeId:          info.VolumeID,
-					StagingTargetPath: cl.Context.StagingPath,
+				&csi.NodeUnpublishVolumeRequest{
+					VolumeId:   volumeID,
+					TargetPath: cl.Context.TargetPath + "/target",
 				},
 			); err != nil {
-				logger.Printf("warning: NodeUnstageVolume: %s", err)
+				if status.Code(err) != codes.NotFound {
+					Fail(fmt.Sprintf("NodeUnpublishVolume failed: %s", err))
```

Comment:
We can't `Fail` during cleanup while there is still work to do. Instead the code should try to execute all operations, log and/or collect failures, and then in the end fail the test.

Otherwise cleaning up stops early although some other volumes perhaps could be deleted successfully.

Reply:
That's a good point. I created a small logger wrapper that tries to simplify the task.